Add support for referenceTarget.

alice opened the pull request:
#### c638de65783f7bdc50e27c165eac138956d000d6
<pre>
Add support for referenceTarget.
<a href="https://bugs.webkit.org/show_bug.cgi?id=285634">https://bugs.webkit.org/show_bug.cgi?id=285634</a>

Reviewed by Ryosuke Niwa.

Explainer: <a href="https://github.com/WICG/webcomponents/blob/gh-pages/proposals/reference-target-explainer.md">https://github.com/WICG/webcomponents/blob/gh-pages/proposals/reference-target-explainer.md</a>

Reference Target is a feature to enable using IDREF attributes such as for and
aria-labelledby to refer to elements inside a component&apos;s shadow DOM, while
maintaining encapsulation of the internal details of the shadow DOM. The main
goal of this feature is to enable ARIA to work across shadow root boundaries.

* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/aria-labelledby-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/dom-mutation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/label-descendant-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/label-for-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/popovertarget-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/property-reflection-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/reference-target-basics-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::addRelation):
* Source/WebCore/dom/CustomElementDefaultARIA.cpp:
(WebCore::CustomElementDefaultARIA::elementForAttribute const):
(WebCore::CustomElementDefaultARIA::elementsForAttribute const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::elementForAttributeInternal const):
(WebCore::Element::getElementAttributeForBindings const):
(WebCore::Element::elementsArrayForAttributeInternal const):
(WebCore::Element::getElementsArrayAttributeForBindings const):
(WebCore::Element::resolveReferenceTarget const):
(WebCore::Element::retargetReferenceTargetForBindings const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::ShadowRoot):
(WebCore::ShadowRoot::setReferenceTarget):
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/dom/ShadowRoot.idl:
* Source/WebCore/dom/ShadowRootInit.h:
* Source/WebCore/dom/ShadowRootInit.idl:
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::elementByIdResolvingReferenceTarget const):
* Source/WebCore/dom/TreeScope.h:
* Source/WebCore/html/FormAssociatedElement.cpp:
(WebCore::FormAssociatedElement::formForBindings const):
* Source/WebCore/html/FormListedElement.cpp:
(WebCore::findAssociatedForm):
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLFormElement.cpp:
(WebCore::HTMLFormElement::formElementIndex):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::list const):
* Source/WebCore/html/HTMLLabelElement.cpp:
(WebCore::HTMLLabelElement::control const):
(WebCore::HTMLLabelElement::controlForBindings const):
(WebCore::HTMLLabelElement::formForBindings const):
* Source/WebCore/html/HTMLLegendElement.cpp:
(WebCore::HTMLLegendElement::formForBindings const):
* Source/WebCore/html/HTMLOptionElement.cpp:
(WebCore::HTMLOptionElement::formForBindings const):
* Source/WebCore/html/HTMLTemplateElement.idl:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::HTMLConstructionSite::insertHTMLTemplateElement):

Canonical link: <a href="https://commits.webkit.org/290529@main">https://commits.webkit.org/290529@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c974090fa851068ff9dd960d07c2668a6d9113b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90321 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9850 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45245 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95325 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41097 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92373 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10238 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18167 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69523 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27109 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93322 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7836 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81933 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49886 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7563 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36301 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40228 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83125 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77889 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37370 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97149 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89099 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17509 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12872 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78519 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17766 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77765 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77732 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19195 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22191 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20800 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/10768 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17519 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22846 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/111590 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17260 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26709 "Failed to checkout and rebase branch from PR 39742") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20712 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19044 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alice commented on Source/WebCore/dom/ShadowRoot.h:
```diff
@@ -181,6 +186,8 @@ class ShadowRoot final : public DocumentFragment, public TreeScope {
     std::unique_ptr<SlotAssignment> m_slotAssignment;
     mutable std::optional<PartMappings> m_partMappings;
+
+    AtomString m_referenceTarget;
```
This is potentially controversial to add to every shadow root for a feature likely to be relatively seldom used. Is there any good alternative?

rniwa replied:
We could consider adding it to `NodeRareData` but I doubt that one extra pointer will matter in the grand scheme of things.

alice replied:
Ok :)

keithamus commented on LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/anchor.html:
```diff
@@ -0,0 +1,109 @@
+<!DOCTYPE HTML><!-- webkit-test-runner [ ShadowRootReferenceTargetEnabled=true ] -->
```
Generally WebKit tends to sync WPTs in a separate PR, and track the import with a separate bug. 

I don't know if you imported these using the script, but if not, running something like:

```
./Tools/Scripts/import-w3c-tests -t web-platform-tests/shadow-dom/reference-target/
```

and check the end of the import logs which will give you a message to copy/paste into the commit message.

alice replied:
Ah thanks - I did indeed run the import script, but I didn't realise it was meant to be a separate PR.

Do they usually land after the code that makes them pass?

lukewarlow replied:
Generally I import them with expectation files showing the failures. Get that merged and then in the implementation PR reset those expectations to show them now passing.

alice replied:
Thank you!

alice replied:
I've made a separate PR for the WPT tests: https://github.com/WebKit/WebKit/pull/39801

Once that lands, I'll rebase this PR.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/anchor.html:
```diff
@@ -1,4 +1,4 @@
-<!DOCTYPE HTML>
+<!DOCTYPE HTML><!-- webkit-test-runner [ ShadowRootReferenceTargetEnabled=true ] -->
```
We usually don't add those to our WPTs, since we want those to be in sync with upstream. Instead I would just mark `ShadowRootReferenceTargetEnabled` as `testable` in `UnifiedWebPreferences.yaml` so the pref gets enabled on WebKit's test infrastructure.

alice replied:
Understood, thanks!

nt1m commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -167,6 +167,12 @@ static bool nodeAndRendererAreValid(Node* node)
 
+static RefPtr<Element> getElementByIdWithReferenceTarget(const TreeScope& treeScope, const AtomString& elementId)
+{
+    RefPtr<Element> element = treeScope.getElementById(elementId);
```
```suggestion
    RefPtr element = treeScope.getElementById(elementId);
```

WebKit style usually omits types

alice replied:
Thanks, would you be able to help me understand the reasoning behind that?

rniwa replied:
Because specifying `Element` here repeats the information already conveyed by the function name "getElementById".

nt1m commented on Source/WebCore/dom/CustomElementDefaultARIA.cpp:
```diff
@@ -91,8 +91,10 @@ RefPtr<Element> CustomElementDefaultARIA::elementForAttribute(const Element& thi
-        if (thisElement.isInTreeScope())
-            result = thisElement.treeScope().getElementById(stringValue);
+        if (thisElement.isInTreeScope()) {
+            RefPtr<Element> elementForId = thisElement.treeScope().getElementById(stringValue);
```
```suggestion
            RefPtr elementForId = thisElement.treeScope().getElementById(stringValue);
```

nt1m commented on Source/WebCore/dom/CustomElementDefaultARIA.cpp:
```diff
@@ -118,7 +120,8 @@ Vector<Ref<Element>> CustomElementDefaultARIA::elementsForAttribute(const Elemen
             SpaceSplitString idList { stringValue, SpaceSplitString::ShouldFoldCase::No };
             result = WTF::compactMap(idList, [&](auto& id) {
-                return thisElement.treeScope().getElementById(id);
+                RefPtr<Element> elementForId = thisElement.treeScope().getElementById(id);
```
```suggestion
                RefPtr elementForId = thisElement.treeScope().getElementById(id);
```

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3267,6 +3324,27 @@ RefPtr<ShadowRoot> Element::shadowRootForBindings(JSC::JSGlobalObject& lexicalGl
+    if (!document().settings().shadowRootReferenceTargetEnabled())
+        return this;
+
+    RefPtr<const Element> element = this;
```
```suggestion
    RefPtr element = this;
```

nt1m commented on Source/WebCore/html/HTMLLabelElement.cpp:
```diff
@@ -73,18 +73,38 @@ Ref<HTMLLabelElement> HTMLLabelElement::create(Document& document)
-        for (const auto& labelableElement : descendantsOfType<HTMLElement>(*this)) {
-            if (labelableElement.isLabelable())
-                return const_cast<HTMLElement*>(&labelableElement);
+        for (const HTMLElement& descendant : descendantsOfType<HTMLElement>(*this)) {
```
```suggestion
        for (const auto& descendant : descendantsOfType<HTMLElement>(*this)) {
```

alice replied:
Similarly, would you be able to help me understand the reasoning behind preferring less type information here?

rniwa replied:
Ditto. Here, `descendantsOfType<HTMLElement>` tells us exactly what type the iterator will iterate over. There is no point in repeating that information again in the variable declaration.

nt1m commented on Source/WebCore/html/HTMLLabelElement.cpp:
```diff
@@ -73,18 +73,38 @@ Ref<HTMLLabelElement> HTMLLabelElement::create(Document& document)
+            return nullptr;
+
+        Ref<Node> retargeted = treeScope().retargetToScope(*control);
+        return downcast<HTMLElement>(retargeted);
```
```suggestion
        RefPtr control = this->control();
        if (!control)
            return nullptr;

        Ref retargeted = treeScope().retargetToScope(*control);
        return downcast<HTMLElement>(retargeted);
```

alice reviewed (commented):
As discussed elsewhere, I'll work on these style nits while splitting this into more granular PRs.

rniwa commented on Source/WebCore/dom/Element.h:
```diff
@@ -409,10 +409,13 @@ class Element : public ContainerNode {
     inline ShadowRoot* shadowRoot() const; // Defined in ElementRareData.h
     RefPtr<ShadowRoot> shadowRootForBindings(JSC::JSGlobalObject&) const;
+    RefPtr<Element> deepShadowRootReferenceTargetOrSelf();
+    RefPtr<const Element> deepShadowRootReferenceTargetOrSelf() const;
```
Oh, we typically just addd `RefPtr<Element> deepShadowRootReferenceTargetOrSelf() const;`

alice replied:
So even if it's returning `this`, it should cast away constness?

alice replied:
I've combined these into one method which casts away constness. Let me know if that's not what you had in mind.

rniwa commented on Source/WebCore/html/HTMLTemplateElement.idl:
```diff
@@ -39,4 +39,5 @@
     [CEReactions=NotNeeded, Reflect=shadowrootclonable] attribute boolean shadowRootClonable;
     [EnabledBySetting=DeclarativeShadowRootsSerializerAPIsEnabled, CEReactions=NotNeeded, Reflect=shadowrootserializable] attribute boolean shadowRootSerializable;
     [EnabledBySetting=ScopedCustomElementRegistryEnabled, CEReactions=NotNeeded, Reflect=shadowrootcustomelements] attribute boolean shadowRootCustomElements;
+    [CEReactions=NotNeeded, EnabledBySetting=ShadowRootReferenceTargetEnabled] attribute [AtomString] DOMString shadowRootReferenceTarget;
```
Why not just add `Reflect=shadowrootreferencetarget` instead of writing setter & getter manually?

alice replied:
Done.

rniwa commented on Source/WebCore/dom/ShadowRoot.cpp:
```diff
@@ -466,4 +467,30 @@ Vector<RefPtr<WebAnimation>> ShadowRoot::getAnimations()
 
+bool ShadowRoot::hasReferenceTarget() const
+{
+    return !m_referenceTarget.isNull();
```
It seems that we could inline this code?

alice replied:
Done.

rniwa commented on Source/WebCore/dom/ShadowRoot.cpp:
```diff
@@ -466,4 +467,30 @@ Vector<RefPtr<WebAnimation>> ShadowRoot::getAnimations()
+
+const AtomString& ShadowRoot::referenceTarget() const
+{
+    return m_referenceTarget;
```
Ditto.

alice replied:
Done.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3296,6 +3328,26 @@ RefPtr<ShadowRoot> Element::shadowRootForBindings(JSC::JSGlobalObject& lexicalGl
+    return const_cast<Element*>(const_cast<const Element*>(this)->deepShadowRootReferenceTargetOrSelf().get());
+}
+
+RefPtr<const Element> Element::deepShadowRootReferenceTargetOrSelf() const
```
I don't think "deep" or "OrSelf" is adding much value here. I'd just call this simply `referenceTarget()` or maybe `referenceTargetElement()`.

alice replied:
I've renamed this to `resolveReferenceTarget()`, since it's not quite the same as what I'd think of as the "reference target element": I'd consider the latter to be the element in this element's shadow root with an ID matching the reference target string, but this can be `this` if there's no reference target, or an element in a deeper shadow root if the reference target element itself has a shadow root with a reference target.

`resolveReferenceTarget()` matches the language we've tentatively sketched out to use in the spec: https://whatpr.org/html/10995/common-microsyntaxes.html#resolve-the-reference-target

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3296,6 +3328,26 @@ RefPtr<ShadowRoot> Element::shadowRootForBindings(JSC::JSGlobalObject& lexicalGl
+    RefPtr shadow = shadowRoot();
+
+    while (shadow && shadow->hasReferenceTarget()) {
+        element = shadow->referenceTargetElementOrHost();
```
It's odd that this function falls back to the host when reference target isn't set after we've checked that the shadow root has a reference target. Why not just add a member function which returns `referenceTargetElement` or `nullptr`?

alice replied:
Done.

rniwa commented on Source/WebCore/dom/ShadowRoot.cpp:
```diff
@@ -466,4 +467,30 @@ Vector<RefPtr<WebAnimation>> ShadowRoot::getAnimations()
+    m_referenceTarget = referenceTarget;
+}
+
+RefPtr<Element> ShadowRoot::referenceTargetElementOrHost() const
```
I think it would be simpler to have this member function return nullptr when the reference target isn't set.
When this whole function could be inlined in the header file like this:
```cpp
RefPtr<Element> ShadowRoot::referenceTargetElement() const
{
    return m_referenceTarget ? getElementById(m_referenceTarget) : nullptr;
}
```

alice replied:
Hm, I swear I had a reason to do this at some point, but it seems like whatever it was became irrelevant in the meantime. Done.

rniwa commented on Source/WebCore/dom/CustomElementDefaultARIA.cpp:
```diff
@@ -91,8 +91,10 @@ RefPtr<Element> CustomElementDefaultARIA::elementForAttribute(const Element& thi
-            result = thisElement.treeScope().getElementById(stringValue);
+        if (thisElement.isInTreeScope()) {
+            RefPtr elementForId = thisElement.treeScope().getElementById(stringValue);
+            result = elementForId ? elementForId->deepShadowRootReferenceTargetOrSelf() : nullptr;
```
This code is duplicated in getElementByIdWithReferenceTarget in AXObjectCache.cpp.
Maybe we should add a helper function to TreeScope like `TreeScope::elementByIdRespectingReferenceTarget` or something?

alice replied:
Done.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2359,6 +2359,7 @@ static RefPtr<Element> getElementByIdIncludingDisconnected(const Element& startE
 
 RefPtr<Element> Element::elementForAttributeInternal(const QualifiedName& attributeName) const
 {
+    RefPtr<Element> element = nullptr;
```
We don't need to initialize RefPtr to nullptr explicitly like this. It's implicitly initialized by its constructor.

alice replied:
Done.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2436,18 +2444,40 @@ std::optional<Vector<Ref<Element>>> Element::elementsArrayForAttributeInternal(c
+        return std::nullopt;
+
+    if (document().settings().shadowRootReferenceTargetEnabled()) {
+        return compactMap(elements.value(), [&](Ref<Element>& element) -> std::optional<Ref<Element>> {
```
I don't think we need to use compactMap here since we're not filtering content.  Just use a regular map instead and you can get rid of the return type: `-> std::optional<Ref<Element>>`.

alice replied:
Good point, done.

rniwa commented on Source/WebCore/html/FormAssociatedElement.cpp:
```diff
@@ -37,6 +37,12 @@ FormAssociatedElement::FormAssociatedElement(HTMLFormElement* form)
 HTMLFormElement* FormAssociatedElement::formForBindings() const
 {
+    if (asHTMLElement().document().settings().shadowRootReferenceTargetEnabled()) {
+        RefPtr form = this->form();
```
Looks like this code is getting duplicated in 4-5 different places.
Can we add a helper function to share the code?

alice replied:
Done.

rniwa commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -1885,7 +1885,15 @@ Vector<Color> HTMLInputElement::suggestedColors() const
+    if (!list)
+        return nullptr;
+
+    if (document().settings().shadowRootReferenceTargetEnabled() && &list->treeScope() != &treeScope())
```
Maybe something like: `elementForBindingsRespectingReferenceTarget(list)`?
where the function is defined like:
```cpp
bool Element::elementForBindingsRespectingReferenceTarget(const Element& element)
{
    if (document().settings().shadowRootReferenceTargetEnabled() && &element.treeScope() != &treeScope())
        return nullptr;
    return element;
}

alice replied:
I named it `retargetForBindings()` - that's what it's doing in theory, it was just shortcutting in these cases before https://github.com/WICG/webcomponents/issues/1072 was resolved. This means we can be consistent across all the "for bindings" methods, and I've added a TODO to remove the downcasts once the WPT has been updated.

annevk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2368,16 +2369,19 @@ RefPtr<Element> Element::elementForAttributeInternal(const QualifiedName& attrib
     }
 
-    return nullptr;
+    if (element == nullptr)
```
```suggestion
    if (!element)
```

alice replied:
Done.

annevk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2436,18 +2444,40 @@ std::optional<Vector<Ref<Element>>> Element::elementsArrayForAttributeInternal(c
     }
 
-    return std::nullopt;
+    if (!elements.has_value())
```
```suggestion
    if (!elements)
```

alice replied:
Done.

annevk commented on Source/WebCore/html/HTMLLabelElement.cpp:
```diff
@@ -73,14 +73,20 @@ Ref<HTMLLabelElement> HTMLLabelElement::create(Document& document)
+    if (!hasAttributeWithoutSynchronization(forAttr)) {
         // Search the children and descendants of the label element for a form element.
         // per http://dev.w3.org/html5/spec/Overview.html#the-label-element
         // the form element must be "labelable form-associated element".
```
This should be updated / removed. Maybe just link the latest version of the algorithm which will in due course be updated to say what we have here?

alice replied:
Done.

annevk commented on Source/WebCore/html/HTMLLabelElement.cpp:
```diff
@@ -103,6 +117,12 @@ HTMLFormElement* HTMLLabelElement::form() const
 HTMLFormElement* HTMLLabelElement::formForBindings() const
 {
+    if (document().settings().shadowRootReferenceTargetEnabled()) {
+        auto* form = this->form();
```
Why not RefPtr?

alice replied:
It should have been, but this code is gone now anyway.

annevk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3260,11 +3290,12 @@ ExceptionOr<ShadowRoot&> Element::attachShadow(const ShadowRootInit& init, Custo
         WTFMove(registry), scopedRegistry);
     if (registryKind == CustomElementRegistryKind::Null)
         shadow->setUsesNullCustomElementRegistry(); // Set this flag for Element::insertedIntoAncestor.
+    shadow->setReferenceTarget(AtomString(init.referenceTarget));
```
Why is this not part of `ShadowRoot::create` like the other init options?

alice replied:
I was hesitant to add a param to `ShadowRoot::create` and the constructor for a currently-experimental feature, but I guess that ship has sailed when the param has already been threaded all the way through to here...

annevk reviewed (commented):
So nice to see progress on this feature!

alice reviewed (commented):
I've made a separate commit to address the review comments, but naturally I'll squash the commits once it's ready to land.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3405,6 +3457,7 @@ CustomElementRegistry* Element::customElementRegistry() const
     return CustomElementRegistry::registryForElement(*this);
 }
 
+
```
Maybe revert this blank line addition?

alice replied:
Done.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3296,6 +3320,34 @@ RefPtr<ShadowRoot> Element::shadowRootForBindings(JSC::JSGlobalObject& lexicalGl
+    return const_cast<Element*>(element.get());
+}
+
+RefPtr<Element> Element::retargetForBindings(RefPtr<Element> element) const
```
I'd call this retargetReferenceTargetForBindings instead since retargeting only happens when reference target is enabled.

alice replied:
Done.

rniwa commented on Source/WebCore/html/FormAssociatedElement.cpp:
```diff
@@ -37,7 +37,8 @@ FormAssociatedElement::FormAssociatedElement(HTMLFormElement* form)
 HTMLFormElement* FormAssociatedElement::formForBindings() const
 {
-    return form();
+    // TODO: The downcast should be unnecessary, but the WPT was written before https://github.com/WICG/webcomponents/issues/1072 was resolved. Update once the WPT has been updated.
```
Please use FIXME instead.

alice replied:
Done.

rniwa commented on Source/WebCore/html/FormListedElement.cpp:
```diff
@@ -108,7 +108,8 @@ static RefPtr<HTMLFormElement> findAssociatedForm(const HTMLElement& element, HT
                 return nullptr;
             if (&element.traverseToRootNode() == &element.treeScope().rootNode()) {
-                ASSERT(&element.traverseToRootNode() == &newFormCandidate->traverseToRootNode());
+                ASSERT((element.document().settings().shadowRootReferenceTargetEnabled() && &element.traverseToRootNode() == &(element.treeScope().retargetToScope(*newFormCandidate))->treeScope().rootNode())
```
This assertion is really long and hard to parse. I think we can use if statement here instead:
```cpp
if (element.document().settings().shadowRootReferenceTargetEnabled())
    ASSERT(&element.traverseToRootNode() == &(element.treeScope().retargetToScope(*newFormCandidate))->treeScope().rootNode()));
else
    ASSERT(&element.traverseToRootNode() == &newFormCandidate->traverseToRootNode());
```

alice replied:
Nice, done.

rniwa commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -1885,7 +1885,8 @@ Vector<Color> HTMLInputElement::suggestedColors() const
 RefPtr<HTMLElement> HTMLInputElement::list() const
 {
-    return dataList();
+    // TODO: the downcast should be unnecessary, but the WPT was written before https://github.com/WICG/webcomponents/issues/1072 was resolved. Update once the WPT has been updated.
```
Please use FIXME instead.

rniwa commented on Source/WebCore/html/HTMLLabelElement.cpp:
```diff
@@ -103,7 +107,8 @@ HTMLFormElement* HTMLLabelElement::form() const
 HTMLFormElement* HTMLLabelElement::formForBindings() const
 {
-    return form();
+    // TODO: The downcast should be unnecessary, but the WPT was written before https://github.com/WICG/webcomponents/issues/1072 was resolved. Update once the WPT has been updated.
```
Please use FIXME instead.

alice replied:
Done.

rniwa commented on Source/WebCore/html/HTMLLegendElement.cpp:
```diff
@@ -58,7 +58,8 @@ HTMLFormElement* HTMLLegendElement::form() const
 {
-    return form();
+    // TODO: the downcast should be unnecessary, but the WPT was written before https://github.com/WICG/webcomponents/issues/1072 was resolved. Update once the WPT has been updated.
+    return dynamicDowncast<HTMLFormElement>(retargetForBindings(form())).get();
```
Please use FIXME instead.

alice replied:
Done.

rniwa commented on Source/WebCore/html/HTMLOptionElement.cpp:
```diff
@@ -147,7 +147,8 @@ HTMLFormElement* HTMLOptionElement::form() const
 {
-    return form();
+    // TODO: the downcast should be unnecessary, but the WPT was written before https://github.com/WICG/webcomponents/issues/1072 was resolved. Update once the WPT has been updated.
+    return dynamicDowncast<HTMLFormElement>(retargetForBindings(form())).get();
```
Please use FIXME instead.

alice replied:
Done.

rniwa reviewed (approved):
r=me

alice committed:
Add support for referenceTarget.
https://bugs.webkit.org/show_bug.cgi?id=285634

Reviewed by Ryosuke Niwa.

Explainer: https://github.com/WICG/webcomponents/blob/gh-pages/proposals/reference-target-explainer.md

Reference Target is a feature to enable using IDREF attributes such as for and
aria-labelledby to refer to elements inside a component's shadow DOM, while
maintaining encapsulation of the internal details of the shadow DOM. The main
goal of this feature is to enable ARIA to work across shadow root boundaries.

* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/aria-labelledby-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/dom-mutation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/label-descendant-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/label-for-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/popovertarget-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/property-reflection-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/reference-target/tentative/reference-target-basics-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::addRelation):
* Source/WebCore/dom/CustomElementDefaultARIA.cpp:
(WebCore::CustomElementDefaultARIA::elementForAttribute const):
(WebCore::CustomElementDefaultARIA::elementsForAttribute const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::elementForAttributeInternal const):
(WebCore::Element::getElementAttributeForBindings const):
(WebCore::Element::elementsArrayForAttributeInternal const):
(WebCore::Element::getElementsArrayAttributeForBindings const):
(WebCore::Element::resolveReferenceTarget const):
(WebCore::Element::retargetReferenceTargetForBindings const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::ShadowRoot):
(WebCore::ShadowRoot::setReferenceTarget):
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/dom/ShadowRoot.idl:
* Source/WebCore/dom/ShadowRootInit.h:
* Source/WebCore/dom/ShadowRootInit.idl:
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::elementByIdResolvingReferenceTarget const):
* Source/WebCore/dom/TreeScope.h:
* Source/WebCore/html/FormAssociatedElement.cpp:
(WebCore::FormAssociatedElement::formForBindings const):
* Source/WebCore/html/FormListedElement.cpp:
(WebCore::findAssociatedForm):
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLFormElement.cpp:
(WebCore::HTMLFormElement::formElementIndex):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::list const):
* Source/WebCore/html/HTMLLabelElement.cpp:
(WebCore::HTMLLabelElement::control const):
(WebCore::HTMLLabelElement::controlForBindings const):
(WebCore::HTMLLabelElement::formForBindings const):
* Source/WebCore/html/HTMLLegendElement.cpp:
(WebCore::HTMLLegendElement::formForBindings const):
* Source/WebCore/html/HTMLOptionElement.cpp:
(WebCore::HTMLOptionElement::formForBindings const):
* Source/WebCore/html/HTMLTemplateElement.idl:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::HTMLConstructionSite::insertHTMLTemplateElement):

Canonical link: https://commits.webkit.org/290529@main
